README.md
# cdk8s-mongo-sts 

![Release](https://github.com/opencdk8s/cdk8s-mongo-sts/workflows/Release/badge.svg?branch=development)
[![npm version](https://badge.fury.io/js/cdk8s-mongo-sts.svg)](https://badge.fury.io/js/cdk8s-mongo-sts)
[![PyPI version](https://badge.fury.io/py/cdk8s-mongo-sts.svg)](https://badge.fury.io/py/cdk8s-mongo-sts)
![npm](https://img.shields.io/npm/dt/cdk8s-mongo-sts?label=npm&color=green) 
![PyPi](https://img.shields.io/pypi/dm/cdk8s-mongo-sts?label=pypi&color=green)

Create a Replicated, Password protected MongoDB Statefulset on Kubernetes, powered by the [cdk8s project](https://cdk8s.io) 🚀

## Disclaimer 

This construct is under heavy development, and breaking changes will be introduced very often. Please don't forget to version lock your code if you are using this construct.

## Overview

**cdk8s-mongo-sts** is a [cdk8s](https://cdk8s.io) library, and also uses [cvallance/mongo-k8s-sidecar](https://github.com/cvallance/mongo-k8s-sidecar) to manage the MongoDB replicaset.

```typescript
import { Construct } from 'constructs';
import { App, Chart, ChartProps } from 'cdk8s';
import { MyMongo } from 'cdk8s-mongo-sts';

export class MyChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = { }) {
    super(scope, id, props);
        new MyMongo(this, 'dev', {
            image: 'mongo',
            namespace: 'databases',
            defaultReplicas: 3,
            volumeSize: '10Gi',
            createStorageClass: true,
            volumeProvisioner: 'kubernetes.io/aws-ebs',
            storageClassName: "io1-slow",
            storageClassParams: {
              type: 'io1',
              fsType: 'ext4',
              iopsPerGB: "10",
            },
            nodeSelectorParams: {
              database: "dev",
            },
        });
    }
}

const app = new App();
new MyChart(app, 'asd');
app.synth();
```

Create a secret for your DB that starts with the same name as your Statefulset with the following keys :

```
username
password
```

See [this](https://kubernetes.io/docs/concepts/configuration/secret/) for documentation on Kubernetes secrets.

Then the Kubernetes manifests created by `cdk8s synth` command will have Kubernetes resources such as `Statefulset`, `Service`, `ClusterRole`, `ClusterRoleBinding`, `ServiceAccount`, and `StorageClass` as follows.

<details>
<summary>manifest.k8s.yaml</summary>

```yaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: io1-slow
parameters:
  fsType: ext4
  type: io1
  iopsPerGB: "10"
provisioner: kubernetes.io/aws-ebs
reclaimPolicy: Retain
---
apiVersion: v1
kind: Service
metadata:
  name: dev
  namespace: databases
spec:
  clusterIP: None
  ports:
    - port: 27017
      targetPort: 27017
  selector:
    db: dev
  type: ClusterIP
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: get-pods-role
  namespace: databases
rules:
  - apiGroups:
      - "*"
    resources:
      - pods
    verbs:
      - list
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: dev
  namespace: databases
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: dev
  namespace: databases
roleRef:
  apiGroup: ""
  kind: ClusterRole
  name: get-pods-role
subjects:
  - kind: ServiceAccount
    name: dev
    namespace: databases
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: dev
  namespace: databases
spec:
  replicas: 3
  selector:
    matchLabels:
      db: dev
  serviceName: dev
  template:
    metadata:
      labels:
        db: dev
    spec:
      containers:
        - env:
            - name: MONGO_SIDECAR_POD_LABELS
              value: db=dev
            - name: KUBE_NAMESPACE
              value: databases
            - name: MONGODB_DATABASE
              value: admin
            - name: MONGODB_USERNAME
              valueFrom:
                secretKeyRef:
                  key: username
                  name: dev
            - name: MONGODB_PASSWORD
              valueFrom:
                secretKeyRef:
                  key: password
                  name: dev
          image: cvallance/mongo-k8s-sidecar
          name: mongo-sidecar
        - args:
            - --replSet
            - rs0
            - --bind_ip
            - 0.0.0.0
            - --dbpath
            - /data/db
            - --oplogSize
            - "128"
          env:
            - name: MONGO_INITDB_ROOT_USERNAME
              valueFrom:
                secretKeyRef:
                  key: username
                  name: dev
            - name: MONGO_INITDB_ROOT_PASSWORD
              valueFrom:
                secretKeyRef:
                  key: password
                  name: dev
          image: mongo
          name: dev
          ports:
            - containerPort: 27017
          resources:
            limits:
              cpu: 400m
              memory: 512Mi
            requests:
              cpu: 200m
              memory: 256Mi
          volumeMounts:
            - mountPath: /data/db
              name: dev
      nodeSelector:
        database: dev
      securityContext:
        fsGroup: 999
        runAsGroup: 999
        runAsUser: 999
      serviceAccountName: dev
      terminationGracePeriodSeconds: 10
  volumeClaimTemplates:
    - metadata:
        name: dev
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        storageClassName: io1-slow
```

</details>

## Installation

### TypeScript

Use `npm` or `yarn` to install.

```shell
$ npm install -s cdk8s-mongo-sts
```

or

```shell
$ yarn add cdk8s-mongo-sts
```

### Python

```shell
$ pip install cdk8s-mongo-sts
```

## Contribution

1. Fork ([https://github.com/Hunter-Thompson/cdk8s-mongo-sts/fork](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/fork))
2. Bootstrap the repo:
  
    ```bash
    npx projen   # generates package.json 
    yarn install # installs dependencies
    ```
3. Development scripts:
   |Command|Description
   |-|-
   |`yarn compile`|Compiles typescript => javascript
   |`yarn watch`|Watch & compile
   |`yarn test`|Run unit test & linter through jest
   |`yarn test -u`|Update jest snapshots
   |`yarn run package`|Creates a `dist` with packages for all languages.
   |`yarn build`|Compile + test + package
   |`yarn bump`|Bump version (with changelog) based on [conventional commits]
   |`yarn release`|Bump + push to `master`
4. Create a feature branch
5. Commit your changes
6. Rebase your local changes against the master branch
7. Create a new Pull Request (use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) for the title please)

## Licence

[Apache License, Version 2.0](./LICENSE)

## Author

[Hunter-Thompson](https://github.com/Hunter-Thompson)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.11](https://github.com/opencdk8s/cdk8s-mongo-sts/compare/v0.0.10...v0.0.11) (2021-06-21)

### [0.0.10](https://github.com/opencdk8s/cdk8s-mongo-sts/compare/v0.0.9...v0.0.10) (2021-06-21)

### [0.0.9](https://github.com/opencdk8s/cdk8s-mongo-sts/compare/v0.0.8...v0.0.9) (2021-06-21)


### Bug Fixes

* constructs dep issue ([#9](https://github.com/opencdk8s/cdk8s-mongo-sts/issues/9)) ([5e3fb37](https://github.com/opencdk8s/cdk8s-mongo-sts/commit/5e3fb374119a25bc8ff981a0d614fd91515abdb1))

### [0.0.8](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/compare/v0.0.7...v0.0.8) (2021-02-08)


### Bug Fixes

* **var:** minor variable fix ([1fb7830](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/1fb783038b559074946c00eaecc20ab0dc7fd202))

### [0.0.7](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/compare/v0.0.6...v0.0.7) (2021-02-08)

### [0.0.6](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/compare/v0.0.5...v0.0.6) (2021-02-07)


### Features

* make node selectors optional ([#8](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/issues/8)) ([074ba52](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/074ba527a8582eb16f10ce9eca67127a6f066fc7))

### [0.0.5](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/compare/v0.0.4...v0.0.5) (2021-02-07)

### [0.0.4](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/compare/v0.0.3...v0.0.4) (2021-02-06)


### Features

* **storage-class:** add option to create storage class ([#5](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/issues/5)) ([65498bb](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/65498bbf53ee96b78c296b974eb2e73a123b9308))

### 0.0.3 (2021-02-06)

### [0.0.2](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/compare/v0.0.1...v0.0.2) (2021-02-06)

### 0.0.1 (2021-02-06)


### Bug Fixes

* .projenrc.js ([df416b0](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/df416b0410f4c8d77adb4bad36168f30fb7b0297))
* .projenrc.js ([f24245f](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/f24245f2a47fdef58841ab93075aa63f37643c58))
* .projenrc.js ([2808a5a](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/2808a5a9e0e80eb338301f5e7d84b83130090e12))
* anti-tamper check ([5aaf344](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/5aaf3446b8e72d9ccbdc05ba4995d102b05564c0))
* tamper checks ([73bde55](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/73bde55c820ba71c15f35b0e0af7d9c0b5c1f968))
* tamper checks ([378e2dc](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/378e2dc808114454758dcedff75de89c28150252))
* yarn.lock file ([9ce54a9](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/9ce54a902d73c489f1608286e645d2828ff0e764))
* yarn.lock file ([b56f244](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/b56f24446144de310920e52f3bbb5313ee3e3532))
* yarn.lock file ([414317b](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/414317b81e0452861521576fa283fb1c65b41bd7))
* yarn.lock file ([d30d43b](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/d30d43b3867b033681d0d025cc6abcbe7cfa6ff5))
* yarn.lock file ([76c1a0c](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/76c1a0cf8a64a87fd0f319b5ea9ca4fe14983b66))
* yarn.lock file ([6cf7d46](https://github.com/Hunter-Thompson/cdk8s-mongo-sts/commit/6cf7d46efc0fc83c42ff1489deeceeebcef6f2cc))


test/hello.test.ts
test('basic', () => {
  const a = 2;
  // expect(a).toEqual(2) this will pass
  expect(a).toEqual(2);
});

.projenrc.js
const { ConstructLibraryCdk8s } = require('projen');

const project = new ConstructLibraryCdk8s({
  author: 'Hunter Thompson',
  authorAddress: 'aatman@auroville.org.in',
  cdk8sVersion: '1.0.0-beta.8',
  defaultReleaseBranch: 'development',
  stability: 'experimental',
  jsiiFqn: 'projen.ConstructLibraryCdk8s',
  name: '@opencdk8s/cdk8s-mongo-sts',
  npmAccess: 'public',
  repositoryUrl: 'https://github.com/opencdk8s/cdk8s-mongo-sts',

  python: {
    distName: 'cdk8s-mongo-sts',
    module: 'cdk8s_mongo_sts',
  },
  peerDeps: ['constructs@^3.3.5'],

  devDeps: [
    'constructs@3.3.5',
    'prettier@^2.2.1',
    'jsii-pacmak@^1.20.1',
    '@commitlint/config-conventional@^11.0.0',
    '@commitlint/cli@^11.0.0',
    'eslint-import-resolver-node@^0.3.4',
    'eslint-import-resolver-typescript@^2.3.0',
    'eslint-plugin-import@^2.22.1',
    'eslint@^7.19.0',
    'jsii-diff@^1.20.1',
    'jsii-docgen@^1.8.36',
    'jsii@^1.20.1',
    'json-schema@^0.3.0',
  ],

  dependabot: false,
  gitignore: ['package.json', 'test/'],
  pullRequestTemplate: false,
  releaseBranches: ['development'],
  codeCov: true,
  releaseEveryCommit: true,
  // npmRegistry: 'npm.pkg.github.com',
  clobber: false,
  readme: true,
});

project.synth();


src/index.ts
import { Construct } from 'constructs';

import * as k8s from './imports/k8s';

export interface STSOptions {
  /**
   * The Kubernetes namespace where this app to be deployed.
   * @default 'default'
   */
  readonly namespace?: string;

  /**
   * The Docker image to use for this app.
   */
  readonly image: string;

  /**
   * Number of replicas.
   * @default 3
   */
  readonly defaultReplicas?: number;

  /**
   * The Volume size of our DB in string, e.g 10Gi, 20Gi
   */
  readonly volumeSize?: string;

  /**
   * Option to create storage class, if enabled, a storage class will be created for the statefulset
   * @default true
   */
  readonly createStorageClass?: boolean;

  /**
   * The storage class to use for our PVC
   * @default 'gp2-expandable'
   */
  readonly storageClassName?: string;

  /**
   * Each StorageClass has a provisioner that determines what volume plugin is used for provisioning PVs. This field must be specified.
   * See [this](https://kubernetes.io/docs/concepts/storage/storage-classes/#provisioner) for Ref
   * @default 'kubernetes.io/aws-ebs'
   */
  readonly volumeProvisioner?: string;

  /**
   * Storage class params
   * @default - { type = gp2, fsType: ext4 }
   */
  readonly storageClassParams?: { [name: string]: string };

  /**
   * nodeSelector params
   * @default - undefined
   */
  readonly nodeSelectorParams?: { [name: string]: string };

  /**
   * Resources requests for the DB.
   * @default - Requests = { CPU = 200m, Mem = 256Mi }, Limits = { CPU = 400m, Mem = 512Mi }
   */
  readonly resources?: ResourceRequirements;
}

export interface ResourceRequirements {
  /**
   * Maximum resources for the web app.
   * @default - CPU = 400m, Mem = 512Mi
   */
  readonly limits?: ResourceQuantity;

  /**
   * Required resources for the web app.
   * @default - CPU = 200m, Mem = 256Mi
   */
  readonly requests?: ResourceQuantity;
}

export interface ResourceQuantity {
  /**
   * @default - no limit
   */
  readonly cpu?: string;

  /**
   * @default - no limit
   */
  readonly memory?: string;
}

/**
 * MongoDB Stateful Set class.
 */

export class MyMongo extends Construct {
  public readonly name: string;
  public readonly namespace: string;

  constructor(scope: Construct, name: string, opts: STSOptions) {
    super(scope, name);

    const volumerequest = {
      storage: k8s.Quantity.fromString(String(opts.volumeSize)),
    };
    const defaultReplicas = opts.defaultReplicas ?? 3;
    const replicas = defaultReplicas;
    const imageName = opts.image;
    const namespace = opts.namespace ?? 'default';
    this.namespace = namespace;
    var storageClassName = opts.storageClassName ?? 'gp2-expandable';
    const volumeProvisioner = opts.volumeProvisioner ?? 'kubernetes.io/aws-ebs';
    const storageClassParams = opts.storageClassParams ?? { type: 'gp2', fsType: 'ext4' };
    const nodeSelectorParams = opts.nodeSelectorParams ?? undefined;
    const resources = {
      limits: convertQuantity(opts.resources?.limits, {
        cpu: '400m',
        memory: '512Mi',
      }),
      requests: convertQuantity(opts.resources?.requests, {
        cpu: '200m',
        memory: '256Mi',
      }),
    };
    const labels = { db: name };

    if (opts.createStorageClass === true) {
      const storageClassOpts: k8s.KubeStorageClassProps = {
        metadata: {
          name: storageClassName,
        },
        provisioner: volumeProvisioner,
        allowVolumeExpansion: true,
        reclaimPolicy: 'Retain',
        parameters: {
          ...storageClassParams,
        },
      };
      const storageclass = new k8s.KubeStorageClass(this, 'storageclass', storageClassOpts);
      this.name = storageclass.name;
      var storageClassName = storageclass.name;
    }

    const serviceOpts: k8s.KubeServiceProps = {
      metadata: {
        namespace: namespace,
        name: name,
      },
      spec: {
        type: 'ClusterIP',
        clusterIP: 'None',
        ports: [{ port: 27017, targetPort: k8s.IntOrString.fromNumber(27017) }],
        selector: labels,
      },
    };
    const svc = new k8s.KubeService(this, 'service', serviceOpts);
    this.name = svc.name;

    const clusterroleOpts: k8s.KubeClusterRoleProps = {
      metadata: {
        namespace: namespace,
        name: 'get-pods-role',
      },
      rules: [
        {
          verbs: ['list'],
          apiGroups: ['*'],
          resources: ['pods'],
        },
      ],
    };
    const clusterrole = new k8s.KubeClusterRole(
      this,
      'clusterrole',
      clusterroleOpts,
    );
    this.name = clusterrole.name;

    const serviceaccountOpts: k8s.KubeServiceAccountProps = {
      metadata: {
        namespace: namespace,
        name: name,
      },
    };
    const sa = new k8s.KubeServiceAccount(
      this,
      'serviceaccount',
      serviceaccountOpts,
    );
    this.name = sa.name;

    const bindingOpts: k8s.KubeClusterRoleBindingProps = {
      roleRef: {
        kind: 'ClusterRole',
        apiGroup: '',
        name: 'get-pods-role',
      },
      metadata: {
        namespace: namespace,
        name: name,
      },
      subjects: [
        {
          kind: 'ServiceAccount',
          name: name,
          namespace: namespace,
        },
      ],
    };
    const clusterrolebinding = new k8s.KubeClusterRoleBinding(
      this,
      'rolebinding',
      bindingOpts,
    );
    this.name = clusterrolebinding.name;

    const pvcProps: k8s.KubePersistentVolumeClaimProps = {
      metadata: {
        name: name,
      },
      spec: {
        accessModes: ['ReadWriteOnce'],
        storageClassName: storageClassName,
        resources: {
          requests: volumerequest,
        },
      },
    };

    const stsOpts: k8s.KubeStatefulSetProps = {
      metadata: {
        namespace: namespace,
        name: name,
      },
      spec: {
        serviceName: name,
        selector: { matchLabels: labels },
        replicas: replicas,
        template: {
          metadata: { labels: labels },
          spec: {
            containers: [
              {
                name: 'mongo-sidecar',
                image: 'cvallance/mongo-k8s-sidecar',
                env: [
                  {
                    name: 'MONGO_SIDECAR_POD_LABELS',
                    value: `db=${name}`,
                  },
                  {
                    name: 'KUBE_NAMESPACE',
                    value: namespace,
                  },
                  {
                    name: 'MONGODB_DATABASE',
                    value: 'admin',
                  },
                  {
                    name: 'MONGODB_USERNAME',
                    valueFrom: {
                      secretKeyRef: {
                        key: 'username',
                        name: name,
                      },
                    },
                  },
                  {
                    name: 'MONGODB_PASSWORD',
                    valueFrom: {
                      secretKeyRef: {
                        key: 'password',
                        name: name,
                      },
                    },
                  },
                ],
              },
              {
                name: name,
                image: imageName,
                env: [
                  {
                    name: 'MONGO_INITDB_ROOT_USERNAME',
                    valueFrom: {
                      secretKeyRef: {
                        key: 'username',
                        name: name,
                      },
                    },
                  },
                  {
                    name: 'MONGO_INITDB_ROOT_PASSWORD',
                    valueFrom: {
                      secretKeyRef: {
                        key: 'password',
                        name: name,
                      },
                    },
                  },
                ],
                resources: resources,
                ports: [
                  {
                    containerPort: 27017,
                  },
                ],
                args: [
                  '--replSet',
                  'rs0',
                  '--bind_ip',
                  '0.0.0.0',
                  '--dbpath',
                  '/data/db',
                  '--oplogSize',
                  '128',
                ],
                volumeMounts: [
                  {
                    name: name,
                    mountPath: '/data/db',
                  },
                ],
              },
            ],
            terminationGracePeriodSeconds: 10,
            serviceAccountName: name,
            nodeSelector: nodeSelectorParams,
            securityContext: {
              fsGroup: 999,
              runAsGroup: 999,
              runAsUser: 999,
            },
          },
        },
        volumeClaimTemplates: [pvcProps],
      },
    };

    const sts = new k8s.KubeStatefulSet(this, 'sts', stsOpts);
    this.name = sts.name;

  }
}

/**
 * Converts a `ResourceQuantity` type to a k8s.Quantity map.
 *
 * If `user` is defined, the values provided there (or lack thereof) will be
 * passed on. This means that if the user, for example, did not specify a value
 * for `cpu`, this value will be omitted from the resource requirements. This is
 * intentional, in case the user intentionally wants to omit a constraint.
 *
 * If `user` is not defined, `defaults` are used.
 */

function convertQuantity(
  user: ResourceQuantity | undefined,
  defaults: { cpu: string; memory: string },
): { [key: string]: k8s.Quantity } {
  // defaults
  if (!user) {
    return {
      cpu: k8s.Quantity.fromString(defaults.cpu),
      memory: k8s.Quantity.fromString(defaults.memory),
    };
  }

  const result: { [key: string]: k8s.Quantity } = {};

  if (user.cpu) {
    result.cpu = k8s.Quantity.fromString(user.cpu);
  }

  if (user.memory) {
    result.memory = k8s.Quantity.fromString(user.memory);
  }

  return result;
}
